system/services/security/PermissionService.cfc

```diff
@@ -225,6 +225,59 @@ component displayName="Admin permissions service" {
 		return groups;
 	}
 
+	public array function listUserGroupsRoles( required string userId ) {
+		var cacheKey = "userGroupsRoles-#arguments.userId#";
+		var fromCache = _getCacheProvider().get( cacheKey );
+		if ( !isNull( local.fromCache ) ) {
+			return fromCache;
+		}
+
+		var groupsRoles = [];
+		var userGroups  = listUserGroups( argumentCollection=arguments );
+
+		if ( arrayLen( userGroups ) ) {
+			var rolesQuery = _getGroupDao().selectData(
+				  filter       = { id=userGroups }
+				, selectFields = [ "roles" ]
+			);
+
+			for ( var q in rolesQuery ) {
+				for ( var role in q.roles ) {
```

Comment:
Should this be `ListToArray( q.roles )`? It is unclear how we're looping over this field from looking at the code.